plugins/optimization-detective/optimization.php

```diff
@@ -112,7 +114,53 @@ static function () use ( $reasons ): void {
 		return;
 	}
 
-	$callback = 'od_optimize_template_output_buffer';
+	$slug    = od_get_url_metrics_slug( od_get_normalized_query_vars() );
+	$post    = OD_URL_Metrics_Post_Type::get_post( $slug );
+	$post_id = $post instanceof WP_Post && $post->ID > 0 ? $post->ID : null;
+
+	$tag_visitor_registry = new OD_Tag_Visitor_Registry();
+
+	/**
+	 * Fires to register tag visitors before walking over the document to perform optimizations.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @param OD_Tag_Visitor_Registry $tag_visitor_registry Tag visitor registry.
+	 */
+	do_action( 'od_register_tag_visitors', $tag_visitor_registry );
+
+	global $wp_the_query;
+	$current_etag     = od_get_current_url_metrics_etag( $tag_visitor_registry, $wp_the_query, od_get_current_theme_template() ); // TODO: Make sure the template is set!!
```

Comment:
I think this may be too early. The `wp` action runs before the template is determined.

Reply:
Indeed. It is too early. The `od_get_current_theme_template()` function returns `null` at this point. It will only be able to return the current template reliably _after_ the `template_include` filter has run, but there is no action that runs between then and the template actually rendering.

Reply:
This works: 8e31a06ce